File: launcher/frontend/src/components/rl/Player.js
/**
 * Player class with physics and controls
 */

export class Player {
  constructor(x, y, width = 32, height = 32) {
    this.x = x
    this.y = y
    this.width = width
    this.height = height
    this.velocityX = 0
    this.velocityY = 0
    this.speed = 5
    this.jumpPower = 15
    this.gravity = 0.8
    this.maxFallSpeed = 20
    this.isOnGround = false
    this.isAlive = true
    this.distance = 0
    this.score = 0
    this.isSprinting = false
    this.isDucking = false
    this.jumpCount = 0  // Track total jumps for metrics
    this.onIcePlatform = false  // Track if on ice (slippery)
  }

  update(platforms, deltaTime = 1) {
    if (!this.isAlive) return

    // Apply gravity with deltaTime for frame-rate independence
    this.velocityY += this.gravity * deltaTime
    if (this.velocityY > this.maxFallSpeed) {
      this.velocityY = this.maxFallSpeed
    }

    // Move horizontally first
    this.x += this.velocityX * deltaTime

    // Check horizontal collisions - platforms are SOLID
    for (const platform of platforms) {
      if (this.isOverlapping(platform)) {
        // Resolve collision by pushing player out
        if (this.velocityX > 0) {
          // Moving right, hit left side of platform
          this.x = platform.x - this.width
        } else if (this.velocityX < 0) {
          // Moving left, hit right side of platform
          this.x = platform.x + platform.width
        }
        this.velocityX = 0
        break
      }
    }

    // Move vertically
    this.y += this.velocityY * deltaTime

    // Check vertical collisions - platforms are SOLID
    this.isOnGround = false
    this.onIcePlatform = false
    for (const platform of platforms) {
      if (this.isOverlapping(platform)) {
        // Resolve collision by pushing player out
        if (this.velocityY > 0) {
          // Moving down, hit top of platform (landing)
          this.y = platform.y - this.height
          this.isOnGround = true
          // Check if platform is ice (slippery)
          if (platform.type === 'ice') {
            this.onIcePlatform = true
          }
        } else if (this.velocityY < 0) {
          // Moving up, hit bottom of platform
          this.y = platform.y + platform.height
        }
        this.velocityY = 0
        break
      }
    }

    // Apply friction (reduced on ice)
    if (this.isOnGround) {
      if (this.onIcePlatform) {
        // Ice: very low friction (0.95 = keeps 95% of velocity)
        this.velocityX *= 0.95
        // Cap velocity on ice to prevent infinite acceleration
        const maxIceVelocity = 15
        if (Math.abs(this.velocityX) > maxIceVelocity) {
          this.velocityX = Math.sign(this.velocityX) * maxIceVelocity
        }
      } else {
        // Normal: high friction (0.7 = keeps 70% of velocity)
        this.velocityX *= 0.7
      }
    }

    // Track distance
    this.distance = Math.max(this.distance, this.x)

    // Die if fall off map
    if (this.y > 1200) {
      this.die()
    }
  }

  // Simple AABB collision detection - checks if player overlaps with platform
  isOverlapping(platform) {
    return (
      this.x < platform.x + platform.width &&
      this.x + this.width > platform.x &&
      this.y < platform.y + platform.height &&
      this.y + this.height > platform.y
    )
  }

  moveLeft() {
    const speed = this.isSprinting ? this.speed * 1.5 : this.speed
    if (this.onIcePlatform) {
      // On ice: add to velocity (acceleration-based for sliding)
      this.velocityX -= speed * 0.3  // 30% acceleration on ice
    } else {
      // Normal: direct control
      this.velocityX = -speed
    }
  }

  moveRight() {
    const speed = this.isSprinting ? this.speed * 1.5 : this.speed
    if (this.onIcePlatform) {
      // On ice: add to velocity (acceleration-based for sliding)
      this.velocityX += speed * 0.3  // 30% acceleration on ice
    } else {
      // Normal: direct control
      this.velocityX = speed
    }
  }

  jump() {
    // Only jump if on ground and not already jumping
    if (this.isOnGround && this.velocityY >= 0) {
      this.velocityY = -this.jumpPower
      this.isOnGround = false
      this.jumpCount++  // Increment jump counter for metrics
    }
  }

  sprint(active) {
    this.isSprinting = active
  }

  duck(active) {
    this.isDucking = active
    // Could reduce height when ducking
  }

  stopMovement() {
    this.velocityX = 0
  }

  die() {
    this.isAlive = false
    this.velocityX = 0
    this.velocityY = 0
  }

  reset(x, y) {
    this.x = x
    this.y = y
    this.velocityX = 0
    this.velocityY = 0
    this.isAlive = true
    this.isOnGround = false
    this.distance = 0
    this.score = 0
    this.jumpCount = 0  // Reset jump counter for new game
  }

  collectCoin(coin) {
    if (!coin.collected) {
      const dx = this.x + this.width / 2 - coin.x
      const dy = this.y + this.height / 2 - coin.y
      const distance = Math.sqrt(dx * dx + dy * dy)

      if (distance < this.width / 2 + coin.radius) {
        coin.collected = true
        this.score += 10
        return true
      }
    }
    return false
  }

  checkEnemyCollision(enemy) {
    return (
      this.x < enemy.x + enemy.width &&
      this.x + this.width > enemy.x &&
      this.y < enemy.y + enemy.height &&
      this.y + this.height > enemy.y
    )
  }

  checkGoalReached(goal) {
    return (
      this.x < goal.x + goal.width &&
      this.x + this.width > goal.x &&
      this.y < goal.y + goal.height &&
      this.y + this.height > goal.y
    )
  }
}

File: RL/backend/training/player.py
"""
Player class with physics and controls
Python translation of launcher/frontend/src/components/rl/Player.js
"""

import math


class Player:
    """
    Player character for the platformer game.
    Handles physics, movement, and collision detection.
    """

    def __init__(self, x, y, width=32, height=32):
        self.x = x
        self.y = y
        self.width = width
        self.height = height
        self.velocityX = 0
        self.velocityY = 0
        self.speed = 5
        self.jumpPower = 15
        self.gravity = 0.8
        self.maxFallSpeed = 20
        self.isOnGround = False
        self.isAlive = True
        self.distance = 0
        self.score = 0
        self.isSprinting = False
        self.isDucking = False
        self.onIcePlatform = False  # Track if on ice (slippery)

    def update(self, platforms, deltaTime=1):
        """
        Update player physics and collision.

        Args:
            platforms: List of platform dictionaries
            deltaTime: Time delta (not used, kept for compatibility)
        """
        if not self.isAlive:
            return

        # Apply gravity
        self.velocityY += self.gravity
        if self.velocityY > self.maxFallSpeed:
            self.velocityY = self.maxFallSpeed

        # Apply velocity
        self.x += self.velocityX
        self.y += self.velocityY

        # Check ground collision
        self.isOnGround = False
        self.onIcePlatform = False
        for platform in platforms:
            if self.checkPlatformCollision(platform):
                self.isOnGround = True
                self.velocityY = 0
                self.y = platform['y'] - self.height
                # Check if platform is ice (slippery)
                if platform.get('type') == 'ice':
                    self.onIcePlatform = True
                break

        # Apply friction (reduced on ice)
        if self.isOnGround:
            if self.onIcePlatform:
                # Ice: very low friction (0.95 = keeps 95% of velocity)
                self.velocityX *= 0.95
                # Cap velocity on ice to prevent infinite acceleration
                max_ice_velocity = 15
                if abs(self.velocityX) > max_ice_velocity:
                    self.velocityX = math.copysign(max_ice_velocity, self.velocityX)
            else:
                # Normal: high friction (0.7 = keeps 70% of velocity)
                self.velocityX *= 0.7

        # Track distance
        self.distance = max(self.distance, self.x)

        # Die if fall off map
        if self.y > 1200:
            self.die()

    def checkPlatformCollision(self, platform):
        """
        Check if player is falling and landing on top of platform.

        Args:
            platform: Platform dictionary with x, y, width, height

        Returns:
            bool: True if collision detected
        """
        # Check if player is falling
        if self.velocityY >= 0:
            playerBottom = self.y + self.height
            platformTop = platform['y']

            # Check if player's feet are near platform top
            if (playerBottom >= platformTop and
                playerBottom <= platformTop + self.velocityY + self.gravity + 5 and
                self.x + self.width > platform['x'] and
                self.x < platform['x'] + platform['width']):
                return True

        return False

    def moveLeft(self):
        """Move player left (or faster if sprinting)"""
        speed = self.speed * 1.5 if self.isSprinting else self.speed
        if self.onIcePlatform:
            # On ice: add to velocity (acceleration-based for sliding)
            self.velocityX -= speed * 0.3  # 30% acceleration on ice
        else:
            # Normal: direct control
            self.velocityX = -speed

    def moveRight(self):
        """Move player right (or faster if sprinting)"""
        speed = self.speed * 1.5 if self.isSprinting else self.speed
        if self.onIcePlatform:
            # On ice: add to velocity (acceleration-based for sliding)
            self.velocityX += speed * 0.3  # 30% acceleration on ice
        else:
            # Normal: direct control
            self.velocityX = speed

    def jump(self):
        """Jump if on ground and not already jumping"""
        if self.isOnGround and self.velocityY >= 0:
            self.velocityY = -self.jumpPower
            self.isOnGround = False

    def sprint(self, active):
        """Set sprint state"""
        self.isSprinting = active

    def duck(self, active):
        """Set duck state (could reduce height when ducking)"""
        self.isDucking = active

    def stopMovement(self):
        """Stop horizontal movement"""
        self.velocityX = 0

    def die(self):
        """Kill the player"""
        self.isAlive = False
        self.velocityX = 0
        self.velocityY = 0

    def reset(self, x, y):
        """Reset player to starting position"""
        self.x = x
        self.y = y
        self.velocityX = 0
        self.velocityY = 0
        self.isAlive = True
        self.isOnGround = False
        self.distance = 0
        self.score = 0

    def collectCoin(self, coin):
        """
        Check if player collected a coin.

        Args:
            coin: Coin dictionary with x, y, radius, collected

        Returns:
            bool: True if coin was collected
        """
        if not coin.get('collected', False):
            dx = self.x + self.width / 2 - coin['x']
            dy = self.y + self.height / 2 - coin['y']
            distance = math.sqrt(dx * dx + dy * dy)

            if distance < self.width / 2 + coin['radius']:
                coin['collected'] = True
                self.score += 10
                return True

        return False

    def checkEnemyCollision(self, enemy):
        """
        Check AABB collision with enemy.

        Args:
            enemy: Enemy dictionary with x, y, width, height

        Returns:
            bool: True if collision detected
        """
        return (
            self.x < enemy['x'] + enemy['width'] and
            self.x + self.width > enemy['x'] and
            self.y < enemy['y'] + enemy['height'] and
            self.y + self.height > enemy['y']
        )

    def checkGoalReached(self, goal):
        """
        Check if player reached the goal.

        Args:
            goal: Goal dictionary with x, y, width, height

        Returns:
            bool: True if goal reached
        """
        return (
            self.x < goal['x'] + goal['width'] and
            self.x + self.width > goal['x'] and
            self.y < goal['y'] + goal['height'] and
            self.y + self.height > goal['y']
        )
